File: Algorithms/515.FindLargestValueInEachTreeRow.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        if not root:
            return result
        cur_level, next_level = [], []
        cur_level.append(root)
        maximum = root.val
        while cur_level:
            node = cur_level.pop(0)
            maximum = max(maximum, node.val)
            if node.left:
                next_level.append(node.left)
            if node.right:
                next_level.append(node.right)
            if not cur_level:
                result.append(maximum)
                if not next_level:
                    break;
                cur_level = next_level
                next_level = []
                maximum = cur_level[0].val
        return result

File: Algorithms/515.FindLargestValueInEachTreeRow.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        List<TreeNode> cur_level = new ArrayList<>();
        List<TreeNode> next_level = new ArrayList<>();
        cur_level.add(root);
        int maximum = root.val;
        while(cur_level.size() != 0) {
            TreeNode node = cur_level.get(0);
            cur_level.remove(0);
            maximum = Math.max(maximum, node.val);
            if (node.left != null) next_level.add(node.left);
            if (node.right != null) next_level.add(node.right);
            if (cur_level.size() == 0) {
                result.add(maximum);
                if (next_level.size() == 0) break;
                cur_level = next_level;
                next_level = new ArrayList<>();
                maximum = cur_level.get(0).val;
            }
        }
        return result;
    }
}
